projects/src/transmitter.cc

```diff
@@ -0,0 +1,157 @@
+
+    // Block USB device until ready
+    while ((SDU1.config->usbp->state != USB_ACTIVE) || (SDU1.state != SDU_READY)) {
+        chThdSleepMilliseconds(10);
+    }
+}
+
+void Transmitter::start(tprio_t priority) {
+    m_thread = chThdCreateStatic(m_wa_transmitter_thread,
+            sizeof(m_wa_transmitter_thread), priority,
+            transmitter_thread_function, this);
+}
+
+BicyclePoseMessage* Transmitter::alloc_pose_message() {
+    void* msg = chPoolAlloc(&m_pose_message_pool);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(msg != nullptr, "Increase transmitter POSE_MESSAGE_POOL_SIZE");
+#endif
+    return static_cast<BicyclePoseMessage*>(msg);
+}
+
+void Transmitter::free_pose_message(BicyclePoseMessage* msg) {
+    chPoolFree(&m_pose_message_pool, reinterpret_cast<void*>(msg));
+}
+
+msg_t Transmitter::transmit_async(BicyclePoseMessage* msg) {
+    msg_t status = chMBPost(&m_message_mailbox, reinterpret_cast<msg_t>(msg), TIME_IMMEDIATE);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(status == MSG_OK, "Increase transmitter MAILBOX_SIZE");
+#endif
+    return status;
+}
+
+SimulationMessage* Transmitter::alloc_simulation_message() {
+    void* msg = chPoolAlloc(&m_simulation_message_pool);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(msg != nullptr, "Increase transmitter SIMULATION_MESSAGE_POOL_SIZE");
+#endif
+    return static_cast<SimulationMessage*>(msg);
+}
+
+void Transmitter::free_simulation_message(SimulationMessage* msg) {
+    chPoolFree(&m_simulation_message_pool, reinterpret_cast<void*>(msg));
+}
+
+msg_t Transmitter::transmit_async(SimulationMessage* msg) {
+    msg_t status = chMBPost(&m_message_mailbox, reinterpret_cast<msg_t>(msg), TIME_IMMEDIATE);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(status == MSG_OK, "Increase transmitter MAILBOX_SIZE");
+#endif
+    return status;
+}
+
+void Transmitter::transmit(SimulationMessage* msg) {
+    if (m_thread != nullptr) {
+        chSysHalt("Synchronization is not implemented and this function is not thread safe");
+    }
+
+    encode_message(reinterpret_cast<msg_t>(msg));
+    transmit_packet();
+    free_simulation_message(msg);
+}
+
+void Transmitter::initialize() {
+    chMBObjectInit(&m_message_mailbox, m_message_mailbox_buffer, MAILBOX_SIZE);
+    chPoolObjectInit(&m_pose_message_pool, sizeof(BicyclePoseMessage), nullptr);
+    chPoolLoadArray(&m_pose_message_pool, static_cast<void*>(m_pose_message_buffer), POSE_MESSAGE_POOL_SIZE);
+    chPoolObjectInit(&m_simulation_message_pool, sizeof(SimulationMessage), nullptr);
+    chPoolLoadArray(&m_simulation_message_pool, static_cast<void*>(m_simulation_message_buffer), SIMULATION_MESSAGE_POOL_SIZE);
+}
+
+void Transmitter::encode_message(msg_t msg) {
+    // For now, we send a simulation message but this needs to be fixed later on.
+    // TODO: Presend protobuf tag. See https://github.com/oliverlee/phobos/issues/181#issuecomment-301825244
+    m_bytes_written = 0;
+    if ((msg % sizeof(stkalign_t)) == 0) {
```

Comment:
Address of memory pool objects must be stack aligned. It the pointer we receive is not stack aligned, it will not point to a valid object in a memory pool